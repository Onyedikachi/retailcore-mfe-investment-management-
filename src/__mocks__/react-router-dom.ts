export const useParams = jest.fn();
export const useLocation = jest.fn();
