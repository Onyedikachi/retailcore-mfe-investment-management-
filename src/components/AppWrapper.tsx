import { Provider } from "react-redux";
import { PersistGate } from "redux-persist/integration/react";
import { persistStore } from "redux-persist";
import { ToastContainer, Zoom } from "react-toastify";
import { store } from "../config/store";

function AppWrapper({ children }) {
  let persistor = persistStore(store());

  return (
    <div data-testid="app">
      <Provider data-testid="store" store={store()}>
        <PersistGate loading={<div>loading...</div>} persistor={persistor}>
          <ToastContainer
            position="top-right"
            autoClose={4000}
            draggable={false}
            // icon={false}
            transition={Zoom}
          />
          {children}
        </PersistGate>
      </Provider>
    </div>
  );
}

export default AppWrapper;
