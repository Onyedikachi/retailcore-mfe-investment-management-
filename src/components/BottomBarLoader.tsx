import React from "react";

const BottomBarLoader = () => {
  return (
    <div
      data-testid="bottom-bar-loader"
      className="w-full  p-2 flex justify-center items-center"
    >
      <div className="spinner-border h-10 w-10 border-t border-danger-500 rounded-full animate-spin"></div>
    </div>
  );
};

export default BottomBarLoader;
