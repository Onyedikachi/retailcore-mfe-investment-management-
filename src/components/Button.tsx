import { forwardRef } from "react";
import clsx from "clsx";
import { BiLoaderAlt } from "react-icons/bi";

type Ref = HTMLButtonElement;

export type ButtonProps = React.DetailedHTMLProps<
  React.ButtonHTMLAttributes<HTMLButtonElement>,
  HTMLButtonElement
> & {
  as?: React.ElementType;
  isLoading?: boolean;
  disabled?: boolean;
  type?: "button" | "submit" | "reset"; // Add the default value for type prop
};

const Button = forwardRef<Ref, ButtonProps>((props, ref) => {
  const {
    as: Component = "button", // Use "button" as default if as prop is not provided
    type = "button",
    className,
    children,
    isLoading,
    disabled,
    ...rest
  } = props;
  const merged = clsx("btn", className);

  return (
    <Component
      ref={ref}
      type={type}
      className={merged}
      {...rest}
      disabled={disabled}
    >
      {children}
      {isLoading && (
        <span>
          <BiLoaderAlt data-testid="loader" className="animate-spin" />
        </span>
      )}
    </Component>
  );
});

Button.displayName = "Button";
export default Button;
