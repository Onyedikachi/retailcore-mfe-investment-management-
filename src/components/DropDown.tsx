// @ts-nocheck
import React, { ReactNode, useEffect, useState } from "react";
import { createPortal } from "react-dom";
import { Menu } from "@headlessui/react";
import {
  FaEye,
  FaEdit,
  FaTrash,
  FaPlayCircle,
  FaBan,
  FaClone,
} from "react-icons/fa";
import { usePopper } from "../hooks/use-popper";

export function handleIcons(value: string) {
  switch (value) {
    case "FaEye":
      return <FaEye className="text-[#3FA2F7]" />;
    case "FaEdit":
      return <FaEdit className="text-[#D4A62F]" />;
    case "FaTrash":
      return <FaTrash className="text-[#DC5A5D]" />;
    case "FaBan":
      return <FaBan className="text-[#DC5A5D]" />;
    case "FaPlayCircle":
      return <FaPlayCircle className="text-[#2FB755]" />;
    case "FaClone":
      return <FaClone className="text-[#085CA5]" />;
  }
}
export default function TestDrop({
  options,
  children,
  handleClick,
}: {
  options: any;
  children: ReactNode;
  handleClick?: (e: any) => void;
}) {
  let [trigger, container] = usePopper({
    placement: "bottom-end",
    strategy: "fixed",
    modifiers: [{ name: "offset", options: { offset: [0, 10] } }],
  });

  return (
    <div className="">
      <Menu as="div" className="relative inline-block text-left">
        <span className="inline-flex rounded-md">
          <Menu.Button
            data-testid="trigger"
            ref={trigger}
            className="inline-flex w-full justify-center px-4 py-2 text-sm "
          >
            {children}
          </Menu.Button>
        </span>

        <Portal>
          <Menu.Items
            ref={container}
            className="shadow-[0px_0px_4px_0px_#00000040] mt-2 min-w-[227px]  divide-y divide-gray-100 rounded-b-lg bg-white z-40"
          >
            <div className="">
              {options?.map((item: { id: any; text: string; icon: string }) => (
                <Menu.Item key={item?.text}>
                  {() => (
                    <button
                      data-testid="menu-button"
                      name={item.text}
                      onClick={() => handleClick(item?.text)}
                      className={` group flex w-full items-center whitespace-nowrap px-4 py-[11px] text-sm text-[#636363] gap-x-3 hover:bg-[#F9E5E5]`}
                    >
                      {handleIcons(item?.icon)} {item?.text}
                    </button>
                  )}
                </Menu.Item>
              ))}
            </div>
          </Menu.Items>
        </Portal>
      </Menu>
    </div>
  );
}

function Portal(props: { children: ReactNode }) {
  let { children } = props;
  let [mounted, setMounted] = useState(false);

  useEffect(() => setMounted(true), []);

  if (!mounted) return null;
  return createPortal(children, document.body);
}
