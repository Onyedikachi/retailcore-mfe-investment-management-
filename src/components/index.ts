export { default as Breadcrumbs } from "./Breadcrumbs";
export { default as DropDown } from "./DropDown";
export { default as Button } from "./Button";
export { default as Loader } from "./Loader";
export { default as MultiRangeSlider } from "./MultiRangeSlider";
export { default as ProcessingStatusSlider } from "./ProcessingStatusSlider";
export { default as TopBar } from "./TopBar";
export { default as StatusCard } from "./StatusCard";
export { default as TableComponent } from "./investment/dashboard/TableComponent";
export { default as QuickLinks } from "./QuickLinks";
export { default as FormStepComponent } from "./FormStepComponent";
export { default as ProductSearch } from "./ProductSearch";
