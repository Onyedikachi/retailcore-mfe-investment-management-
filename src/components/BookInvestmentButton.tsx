import React, { useState } from "react";
import OutsideClickHandler from "react-outside-click-handler";
import { BookInvestmentOptions, IBookInvestmentOptions } from "@app/constants";
import { useNavigate } from "react-router-dom";

type BookInvestmentButton = {
  children?: any;
  disabled?: boolean;
};

export default function BookInvestmentButton({
  children,
  disabled,
}: BookInvestmentButton) {
  const navigate = useNavigate();
  const [isOpen, setIsOpen] = useState(false);

  const goToUrl = (url) => {
    if (disabled) return;
    navigate(url);

    setIsOpen(false);
  };

  return (
    <div>
      <OutsideClickHandler
        onOutsideClick={() => {
          setIsOpen(false);
        }}
      >
        <div className="relative">
          <div onKeyUp={() => {}} onClick={() => setIsOpen(true)}>{children}</div>
          {isOpen && (
            <div className="top-0 absolute w-[286px] z-10 bg-white rounded-b-lg shadow-[0px_0px_4px_0px_rgba(0,0,0,0.25)] pt-[1px]">
              <ul>
                {BookInvestmentOptions?.map((item: IBookInvestmentOptions) => (
                  <li className="group" key={item.title}>
                    <button
                      disabled={disabled}
                      onClick={() => {
                        goToUrl(item.url);
                      }}
                      className={`disabled:opacity-50 disabled:cursor-not-allowed w-full z-10 relative text-base text-[#636363] capitalize pl-6 pr-4 flex justify-between items-center py-3 group-last:rounded-b-lg hover:bg-[#F9E5E5] cursor-pointer`}
                    >
                      {item.title}
                    </button>
                  </li>
                ))}
              </ul>
            </div>
          )}
        </div>
      </OutsideClickHandler>
    </div>
  );
}
