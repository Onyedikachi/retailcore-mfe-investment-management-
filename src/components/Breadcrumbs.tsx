import React from "react";
import { BsArrowLeft } from "react-icons/bs";
import { useNavigate } from "react-router-dom";

interface BreadcrumbItem {
  id: string | number;
  title: string;
  url: string;
}

interface BreadcrumbsProps {
  links: BreadcrumbItem[];
}

export default function Breadcrumbs({
  links,
}: BreadcrumbsProps): React.ReactElement {
  const navigate = useNavigate();
  return (
    <div className="flex items-center gap-x-2" data-testid="breadcrumbs">
      <button
        className="py-1 px-1 h-8 w-8 rounded flex items-center justify-center outline-none bg-transparent hover:bg-[#F9F2F2]"
        onClick={() => navigate(-1)}
        data-testid="back button"
      >
        <BsArrowLeft className="text-[#636363]" />{" "}
      </button>
      <ul className="flex gap-x-8 uppercase items-center">
        {links.map((item: BreadcrumbItem, idx: number) => (
          <li
            key={`${idx.toString()}-index`}
            onMouseOver={() => { }}
            onClick={() => navigate(item.url)}
            data-testid={item.title}
            className={`first-of-type:hover:bg-[#F9F2F2] ${item.url !== "#" ? "hover:bg-[#F9F2F2] cursor-pointer" : ""
              } px-2 py-1 rounded first-of-type:cursor-pointer relative text-[#8F8F8F] text-base font-medium last-of-type:text-[#636363] after:absolute after:content-[''] after:right-[-16px] after:top-1/2 after:translate-y-[-50%] after:w-2 after:h-2 after:border-b-2 after:border-r-2 after:border-[#636363] after:-rotate-45 last-of-type:after:content-none`}
          >
            <span>{item.title}</span>
          </li>
        ))}
      </ul>
    </div>
  );
}
