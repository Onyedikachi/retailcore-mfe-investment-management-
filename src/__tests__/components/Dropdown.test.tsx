// Breadcrumbs.test.jsx

import React from "react";
import { screen, fireEvent, render } from "@testing-library/react";
import {
  FaEye,
  FaEdit,
  FaTrash,
  FaPlayCircle,
  FaBan,
  FaBars,
} from "react-icons/fa";
import { MemoryRouter } from "react-router-dom";
import { DropDown } from "../../components";
import { handleIcons } from "../../components/DropDown";

const options = [
  {
    id: "",
    text: "View",
    icon: "FaEye",
  },
  {
    id: "",
    text: "Modify",
    icon: "FaEdit",
  },
  {
    id: "",
    text: "Delete Request",
    icon: "FaTrash",
  },
];
jest.mock("react-router-dom", () => ({
  BrowserRouter: ({ children }) => <div>{children}</div>,
  Routes: ({ children }) => <div>{children}</div>,
  Route: ({ element }) => element,
}));

// Mock handleClick function
const handleClick = jest.fn();
let mockActive = true;
describe("Dropdown", () => {
  it("renders without crashing", () => {
    render(
      <DropDown
        options={options}
        children={<FaBars className=" text-sterling-red-800" />}
        handleClick={function (e: any): void {
          throw new Error("Function not implemented.");
        }}
      />
    );
  });

  it("matches snapshot", () => {
    const { asFragment } = render(
      <DropDown
        options={options}
        children={<FaBars className=" text-sterling-red-800" />}
        handleClick={function (e: any): void {
          throw new Error("Function not implemented.");
        }}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("selects the right icon", () => {
    render(
      <DropDown
        options={options}
        children={undefined}
        handleClick={function (e: any): void {
          throw new Error("Function not implemented.");
        }}
      />
    );
    const eyeResult = handleIcons("FaEye");
    expect(eyeResult).toStrictEqual(<FaEye className="text-[#3FA2F7]" />);

    const editResult = handleIcons("FaEdit");
    expect(editResult).toStrictEqual(<FaEdit className="text-[#D4A62F]" />);

    const trashResult = handleIcons("FaTrash");
    expect(trashResult).toStrictEqual(<FaTrash className="text-[#DC5A5D]" />);

    const banResult = handleIcons("FaBan");
    expect(banResult).toStrictEqual(<FaBan className="text-[#DC5A5D]" />);

    const playResult = handleIcons("FaPlayCircle");
    expect(playResult).toStrictEqual(
      <FaPlayCircle className="text-[#2FB755]" />
    );
  });

  it("renders right icon when button is clicked", () => {
    const { getAllByTestId, getByTestId } = render(
      <DropDown
        options={options}
        children={<FaBars className=" text-sterling-red-800" />}
        handleClick={handleClick}
      />
    );

    const button = getByTestId("trigger");
    fireEvent.click(button);
    const menus = getAllByTestId("menu-button");
    expect(menus[0]).toHaveTextContent(/^View$/);
    expect(menus[1]).toHaveTextContent(/^Modify$/);
    expect(menus[2]).toHaveTextContent(/^Delete Request$/);

    fireEvent.click(menus[0]);
    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  it("applies active class to the selected option", () => {
    const { getByTestId } = render(
      <DropDown
        options={options}
        children={<FaBars className=" text-sterling-red-800" />}
        handleClick={handleClick}
      />
    );
    const button = getByTestId("trigger");
    fireEvent.click(button);
    // Initially, no option should have the active class
    options?.forEach((option) => {
      const optionButton = screen.getByText(option.text);
      expect(optionButton).not.toHaveClass("bg-[#F9E5E5]");
    });

    // Simulate a click on the second option
    fireEvent.click(screen.getByText("View"));
  });
});
