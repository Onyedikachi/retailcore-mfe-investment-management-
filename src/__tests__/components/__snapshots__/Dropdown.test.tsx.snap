// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown matches snapshot 1`] = `
<DocumentFragment>
  <div
    class=""
  >
    <div
      class="relative inline-block text-left"
      data-headlessui-state=""
    >
      <span
        class="inline-flex rounded-md"
      >
        <button
          aria-expanded="false"
          aria-haspopup="menu"
          class="inline-flex w-full justify-center px-4 py-2 text-sm "
          data-headlessui-state=""
          data-testid="trigger"
          id="headlessui-menu-button-3"
          type="button"
        >
          <svg
            class=" text-sterling-red-800"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M16 132h416c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H16C7.163 60 0 67.163 0 76v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"
            />
          </svg>
        </button>
      </span>
    </div>
  </div>
</DocumentFragment>
`;
